components/script/dom/bindings/buffer_source_types.rs

```diff
@@ -0,0 +1,441 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#![allow(unsafe_code)]
+
+use std::borrow::BorrowMut;
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use std::ptr;
+use std::sync::{Arc, Mutex};
+
+use js::jsapi::{
+    Heap, JSObject, JS_GetArrayBufferViewBuffer, JS_IsArrayBufferViewObject, NewExternalArrayBuffer,
+};
+use js::rust::wrappers::DetachArrayBuffer;
+use js::rust::{CustomAutoRooterGuard, Handle, MutableHandleObject};
+use js::typedarray::{CreateWith, TypedArray, TypedArrayElement, TypedArrayElementCreator};
+
+use crate::script_runtime::JSContext;
+
+/// <https://webidl.spec.whatwg.org/#BufferSource>
+#[allow(dead_code)]
+pub enum BufferSource {
+    Int8Array(Box<Heap<*mut JSObject>>),
+    Int16Array(Box<Heap<*mut JSObject>>),
+    Int32Array(Box<Heap<*mut JSObject>>),
+    Uint8Array(Box<Heap<*mut JSObject>>),
+    Uint16Array(Box<Heap<*mut JSObject>>),
+    Uint32Array(Box<Heap<*mut JSObject>>),
+    Uint8ClampedArray(Box<Heap<*mut JSObject>>),
+    BigInt64Array(Box<Heap<*mut JSObject>>),
+    BigUint64Array(Box<Heap<*mut JSObject>>),
+    Float32Array(Box<Heap<*mut JSObject>>),
+    Float64Array(Box<Heap<*mut JSObject>>),
+    DataView(Box<Heap<*mut JSObject>>),
+    ArrayBuffer(Box<Heap<*mut JSObject>>),
+    Default(Box<Heap<*mut JSObject>>),
+}
+
+pub struct HeapBufferSourceTypes<T> {
```

Comment:
What does the `Types` here refer to? Maybe this should just be called `HeapBufferSource`?

Reply:
I tried to follow spec blindly, indeed `Types` in not needed here

Reply:
I would rename this to `HeapBufferSource` and this file to `buffersource` then to match the naming scheme we use for the bindings currently.